# Regressão Logística I
## Tarefa II

Vamos trabalhar com a mesma base do exercício anterior, mas vamos aprofundar um pouco mais a nossa regressão.

In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

import statsmodels.formula.api as smf

In [2]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.cleveland.data'

df = pd.read_csv(url, 
                 names=['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg',
                        'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'num'])
df['flag_doente'] = (df['num']!=0).astype('int64')
df.head()

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  63.0  1.0  1.0     145.0  233.0  1.0      2.0    150.0    0.0      2.3   
1  67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0      1.5   
2  67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0      2.6   
3  37.0  1.0  3.0     130.0  250.0  0.0      0.0    187.0    0.0      3.5   
4  41.0  0.0  2.0     130.0  204.0  0.0      2.0    172.0    0.0      1.4   

   slope   ca thal  num  flag_doente  
0    3.0  0.0  6.0    0            0  
1    2.0  3.0  3.0    2            1  
2    2.0  2.0  7.0    1            1  
3    3.0  0.0  3.0    0            0  
4    1.0  0.0  3.0    0            0

A descrição das variáveis está recortada abaixo:
- age: idade do paciente em anos
- sex: sexo (1 = male; 0 = female)  
- cp: tipo de dor no peito
  - 1: angina típica
  - 2: angina atípica
  - 3: dor não-angina
  - 4: assintomático
- trestbps: pressão sanguínea em repouso (em mm Hg na admissão ao hospital
- chol: colesterol sérico em mg/dl
- fbs: (açúcar no sangue em jejum > 120 mg/dl) (1 = True; 0 = False)
- restecg: resultados eletrocardiográficos em repouso
  - 0: normal
  - 1: tendo anormalidade da onda ST-T (Inversões de onda T e / ou ST com elevação ou depressão de > 0.05 mV)
  - 2: mostrando hipertrofia ventricular esquerda provável ou definitiva pelos critérios de Estes
- thalach: frequência cardíaca máxima alcançada
- exang: angina induzida por exercício(1 = sim; 0 = não)
- oldpeak = Depressão de ST induzida por exercício em relação ao repouso
- slope: Depressão de ST induzida por exercício em relação ao repouso
  - 1: inclinação ascendente
  - 2: estável
  - 3: inclinação descendente
- ca: número de vasos principais (0-3) coloridos por fluorosopia
- thal: 3 = normal; 6 = defeito corrigido; 7 = defeito reversível
- num: diagnóstico de doença cardíaga (status de doença angiográfica)

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   age          303 non-null    float64
 1   sex          303 non-null    float64
 2   cp           303 non-null    float64
 3   trestbps     303 non-null    float64
 4   chol         303 non-null    float64
 5   fbs          303 non-null    float64
 6   restecg      303 non-null    float64
 7   thalach      303 non-null    float64
 8   exang        303 non-null    float64
 9   oldpeak      303 non-null    float64
 10  slope        303 non-null    float64
 11  ca           303 non-null    object 
 12  thal         303 non-null    object 
 13  num          303 non-null    int64  
 14  flag_doente  303 non-null    int64  
dtypes: float64(11), int64(2), object(2)
memory usage: 35.6+ KB


1. Considere o script que monta a análise bivariada que você fez na tarefa anterior. Transforme esse script em uma função, que deve:
- Ter como parâmetros de entrada:
    - Um *dataframe* contendo os dados a serem avaliados
    - Um *string* contendo o nome da variável resposta
    - Um *string* contendo o nome da variável explicativa
- E deve retornar um *dataframe* com os dados da bivariada. 
**Monte** a mesma bivariada pelo menos três variáveis qualitativas do *data-frame*. Qual delas parece discriminar mais o risco?

In [5]:
df.head()

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  63.0  1.0  1.0     145.0  233.0  1.0      2.0    150.0    0.0      2.3   
1  67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0      1.5   
2  67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0      2.6   
3  37.0  1.0  3.0     130.0  250.0  0.0      0.0    187.0    0.0      3.5   
4  41.0  0.0  2.0     130.0  204.0  0.0      2.0    172.0    0.0      1.4   

   slope   ca thal  num  flag_doente  
0    3.0  0.0  6.0    0            0  
1    2.0  3.0  3.0    2            1  
2    2.0  2.0  7.0    1            1  
3    3.0  0.0  3.0    0            0  
4    1.0  0.0  3.0    0            0

In [6]:
def script_analise_biv(nome_data, resposta, explicativa):
    nome_data = pd.crosstab(df[resposta], df[explicativa], margins=True)
    nome_data["Media_de_doentes"] = nome_data[1]/nome_data["All"]
    nome_data["chance"] = nome_data[1]/nome_data[0]
    nome_data.loc["All", "chance"]
    nome_data["RC_vs_All"] = nome_data["chance"]/nome_data.loc["All", "chance"]
    return nome_data

In [7]:
novo_data = script_analise_biv("primeiro", "sex", "flag_doente")
novo_data

flag_doente    0    1  All  Media_de_doentes    chance  RC_vs_All
sex                                                              
0.0           72   25   97          0.257732  0.347222   0.409672
1.0           92  114  206          0.553398  1.239130   1.461996
All          164  139  303          0.458746  0.847561   1.000000

In [8]:
novo_data1 = script_analise_biv("segundo", "slope", "flag_doente")
novo_data1

flag_doente    0    1  All  Media_de_doentes    chance  RC_vs_All
slope                                                            
1.0          106   36  142          0.253521  0.339623   0.400706
2.0           49   91  140          0.650000  1.857143   2.191161
3.0            9   12   21          0.571429  1.333333   1.573141
All          164  139  303          0.458746  0.847561   1.000000

In [9]:
novo_data2 = script_analise_biv("terceiro", "cp", "flag_doente")
novo_data2

flag_doente    0    1  All  Media_de_doentes    chance  RC_vs_All
cp                                                               
1.0           16    7   23          0.304348  0.437500   0.516187
2.0           41    9   50          0.180000  0.219512   0.258993
3.0           68   18   86          0.209302  0.264706   0.312315
4.0           39  105  144          0.729167  2.692308   3.176536
All          164  139  303          0.458746  0.847561   1.000000

2. Monte uma função semelhante para categorizar variáveis quantitativas contínuas (com muitas categorias) como ```age```.  
    Além dos mesmos parâmetros da função anterior, defina mais um parâmetro como número de categorias que você deseja quebrar. Defina um valor '*default*' de 5 grupos para este parâmetro.  

In [10]:
df.head()

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  63.0  1.0  1.0     145.0  233.0  1.0      2.0    150.0    0.0      2.3   
1  67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0      1.5   
2  67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0      2.6   
3  37.0  1.0  3.0     130.0  250.0  0.0      0.0    187.0    0.0      3.5   
4  41.0  0.0  2.0     130.0  204.0  0.0      2.0    172.0    0.0      1.4   

   slope   ca thal  num  flag_doente  
0    3.0  0.0  6.0    0            0  
1    2.0  3.0  3.0    2            1  
2    2.0  2.0  7.0    1            1  
3    3.0  0.0  3.0    0            0  
4    1.0  0.0  3.0    0            0

In [11]:
def analiseBivariadaQuant(df: pd.DataFrame, variavel_resposta: str, variavel_explicativa: str, n_bins=5) -> pd.DataFrame :

    # Criando uma nova coluna no DataFrame contendo os grupos de variável explicativa com base nos quantis
    df[f'{variavel_explicativa}_grupo'] = pd.qcut(df[variavel_explicativa], q=n_bins)
    # Contando a quantidade de observações em cada grupo
    df[f'{variavel_explicativa}_grupo'].value_counts()

    # Tabela de contingência entre os grupos da variável explicativa e a variável resposta,
    ana_biv = pd.crosstab(df[f'{variavel_explicativa}_grupo'], df[variavel_resposta], margins=True, margins_name='Total')
    # Renomeia as colunas da tabela
    ana_biv.columns = ['cat1', 'cat2', 'Total']
    print(f'Quantidade média de "{variavel_explicativa}":', df[variavel_explicativa].mean())


    # Análises
    # Calculado a média da variável resposta 
    ana_biv[f'media_{variavel_resposta}_por_{variavel_explicativa}'] = ana_biv['cat2'] / ana_biv['Total']
    # Calculando a razão de chances (odds)
    ana_biv['Odds'] = ana_biv['cat2'] / ana_biv['cat1']
    ana_biv['Odds ratio'] = ana_biv['Odds'] / ana_biv.loc[ana_biv.index[0], 'Odds']
    ana_biv['Log (Odds)'] = np.log(ana_biv['Odds'])
    # Calculando o log do odds ratio para cada grupo da variável explicativa
    ana_biv['Log (Odds ratio)'] = np.log(ana_biv['Odds ratio'])
    
    return ana_biv



3. Construa um modelo de regressão logística com as variáveis qualitativas: ```sex + cp +  trestbps``` e com a variável quantitativa ```age```.

**Interprete os parâmetros.**

In [14]:
df.head()

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  63.0  1.0  1.0     145.0  233.0  1.0      2.0    150.0    0.0      2.3   
1  67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0      1.5   
2  67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0      2.6   
3  37.0  1.0  3.0     130.0  250.0  0.0      0.0    187.0    0.0      3.5   
4  41.0  0.0  2.0     130.0  204.0  0.0      2.0    172.0    0.0      1.4   

   slope   ca thal  num  flag_doente  
0    3.0  0.0  6.0    0            0  
1    2.0  3.0  3.0    2            1  
2    2.0  2.0  7.0    1            1  
3    3.0  0.0  3.0    0            0  
4    1.0  0.0  3.0    0            0

In [20]:
import statsmodels.formula.api as sm

In [23]:
reglog = smf.logit(" age ~ cp", data=df).fit()

reglog.summary()

ValueError: endog must be in the unit interval.

4. Avalie o seu modelo quanto a **calibragem**:
- Calcule a probabilidade de evento predita segundo o seu modelo
- Categorize essa probabilidade em G=5 grupos
- Calcule a probabilidade de evento predita média por grupo
- Calcule a taxa de eventos (média da variável indicadora de eventos) por grupo
- Compare graficamente o valor eperado versus observado para a taxa de maus por grupo

5. Avalie o seu modelo quanto a discriminação calculando acurácia, GINI e KS.

6. tente melhorar o modelo obtido, por exemplo inserindo ou removendo variáveis.  
    Avalie as características do seu modelo (calibragem e acurácia).